### ISSUE TITLE
Probability of each genotype under the model?

### ISSUE THREAD

**Issue by rdiaz02** [phillipnicol/OncoBN#3]
Is there code to obtain the probability of each genotype under the model, as given by Figure 1 in the paper? (I am thinking something similar to what Oncotree's, https://cran.r-project.org/web/packages/Oncotree/index.html, `distribution.oncotree` does).  There is function `Likelihood` (in file `fitCPN.R`) which, well, computes the loglikelihood, but it does not directly give the probability of each possible genotype.

**Comment by phillipnicol** [comment]
Yes this can be done... I've just updated the package to include a function `Lik.genotype()` that computes the probability of a specific genotypes under the estimated model. 

The below code chunk shows how the function can be used to compute the probability for all 2^n possible genotypes. It runs of my computer but let me know if it works and if it was what you were looking for. Thanks! 

```
data("example")
out <- fitCPN(example,model="DBN",epsilon=0.05,verbose=FALSE)
n <- 4 #number of mutations
genotypes <- expand.grid(replicate(n, 0:1, simplify = FALSE))
colnames(genotypes) <- colnames(example)
genotypes$Likelihood <- apply(genotypes,1,function(x) {
  Lik.genotype(out,x)
})
```

**Comment by rdiaz02** [comment]
Thanks! Works perfectly.

**Comment by rdiaz02** [comment]
Closing the issue.

**Comment by rdiaz02** [comment]
May I suggest a minor improvement, in the form of a consistency check?

With some synthetic data, I found that two identical fits can give different predictions if the igraph object for graph was constructed differently, even if the adjacency matrix is the same (when rows/columns are ordered identically). As an example, suppose these two models (where I've left only the relevant parts, set epsilon to 0, and named the theta)

```
obnfit1
$graph
IGRAPH 3b16913 DN-- 4 4 -- 
+ > attr: name (v/c)
+ > edges from 3b16913 (vertex names):
[1] WT->A WT->B A ->C B ->C

$theta
  A   B   C 
0.3 0.4 0.6 

$model
[1] "DBN"

$epsilon
[1] 0
```
and
```
obnfit2
$graph
IGRAPH 042a834 DN-- 4 4 -- 
+ > attr: name (v/c)
+ > edges from 042a834 (vertex names):
[1] B ->C WT->B A ->C WT->A

$theta
  A   B   C 
0.3 0.4 0.6 

$model
[1] "DBN"

$epsilon
[1] 0
```
The models are identical. But, when we go into `Lik.genotype` we have this:
```
 G <- t(as.matrix(as_adjacency_matrix(fit$graph)))
```

which gives, for each of the two models respectively

```
G
   WT A B C
WT  0 0 0 0
A   1 0 0 0
B   1 0 0 0
C   0 1 1 0
```
and

``` 
G
   B WT A C
B  0  1 0 0
WT 0  0 0 0
A  0  1 0 0
C  1  0 1 0
```

The two adjacency matrices are identical, if we were to reorder them. 

Just in case, I've added, to the code snippet you provided above, the following check, right after `colnames(genotypes) <- colnames(example)`

```
G <- t(as.matrix(as_adjacency_matrix(fit$graph)))
stopifnot(colnames(genotypes) == colnames(G)[-1])
stopifnot(colnames(G)[1] == "WT")
```

Of course, I understand that the problem is probably only relevant if one messes around with the fit object manually; so if I decide to do so, it is me (not the OncoBN code) who needs to ensure things are consistent (and if bad things happens, well, it is my fault). But this is a cheap check anyway and maybe worth it?

**Comment by rdiaz02** [comment]
This can happen also when I do not mess around with the returned object. The possible problem appears in a run from `fitCPN`. 

```
library(igraph)
library(OncoBN)
data2 <- read.table("data2.txt", header = TRUE)
fit1 <- fitCPN(data = data2, model = "CBN")
```

The output is
``` 
> fit1
$edgelist
[1] "WT" "A"  "A"  "B"  "D"  "C"  "B"  "D" 

$score
[1] -2374

$graph
IGRAPH b5e3548 DN-- 5 4 -- 
+ > attr: name (v/c)
+ > edges from b5e3548 (vertex names):
[1] WT->A A ->B D ->C B ->D

$theta
[1] 0.9921 0.4000 0.5714 0.4000

$model
[1] "CBN"

$epsilon
[1] 0.1389``

```

But there is potential ambiguity there. Notice the order of rows/columns in the adjacency matrix:

```
> t(as.matrix(as_adjacency_matrix(fit1$graph)))
   WT A B D C
WT  0 0 0 0 0
A   1 0 0 0 0
B   0 1 0 0 0
D   0 0 1 0 0
C   0 0 0 1 0
```
Whereas the order of names in the data is A, B, C, D.  And this raises the question: what `thetas` correspond to which gene here?

In other words do we have the thetas in the order of the column names: `A = .9921, B = .4000, C = .5714, D = 0.4000` or do we have them in the order given in the adjacency matrix:  `A = .9921, B = .4000, D = .5714, C = 0.4000`.

From the code of `fitCPN` I think it is the the first (line with ` inferTheta(df, out$graph, model)`), but I am not 100% sure this is always that way.

If the first case, the code in `Lik.genotype` is then wrong, I think.  But if it is the second case, the correct interpretation of the `theta`s would depend on specifics of the behavior of `as_adjacency_matrix` that, as far as I can see, are not documented.

I suggest two no-exclusive ways to solve this issue.

1. For thetas, returned a named vector. 
2. When constructing the graph using igraph, doing it from an adjacency matrix that is ordered in the same way as the column names of the input data frame. So, instead of
```
out$graph <- igraph::make_directed_graph(out$edgelist)
```
a verbose way would be

```
adjm <- igraph::as_adjacency_matrix(igraph::make_directed_graph(out$edgelist))

gn <- setdiff(colnames(df), "WT")
adjm <- adjm[c("WT", gn), c("WT", gn)]
out$graph <- igraph::graph_from_adjacency_matrix(adjm)
```


[data2.txt](https://github.com/phillipnicol/OncoBN/files/8126661/data2.txt)

**Comment by phillipnicol** [comment]
Thanks for pointing this out and sorry for the late response! The returned theta is in the order of the column names, updated it so that theta is now a named vector. I also ensure that the names of the adjacency are in the correct order (in fact WT needs to come first followed by the column names in their original order).

**Comment by rdiaz02** [comment]
Sorry for taking so long to answer. 

I think there is a new minor issue: `Lik.genotype` now has this
```
 order[2:length(order)] <- colnames(df)
```

But `df` is not passed as an argument to `Lik.genotype`. 

Moreover, after the swap in 
```
 ix <- which(order == "WT")
order[c(1, ix)] <- order[c(ix, 1)]
```
I am not sure that `colnames(df)` and the entries in the adjacency matrix would always correspond to genes in the same order.

If the order is guaranteed to be kept, I think a simpler and safer option would be something like:

```
order[2:length(order)] <- rownames(G)[-ix]
```
instead of 

```
 order[2:length(order)] <- colnames(df)
```

Anyway, I think this issue does not arise if the graph is created from an adjacency matrix that is already ordered in the same way as the input data (with the first row/column being "WT"). Then, the previous version of `Lik.genotype` would work just fine.
